File: Data Structures & Algorithms/reorder-linked-list/submission-9.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head.next:
            return

        dummy = ListNode(-1, head)
        slow = dummy
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        curr = slow.next
        slow.next = None
        prev = None

        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp

        p1 = head
        p2 = prev

        while p2:
            tmp = p1.next
            p1.next = p2
            p1 = p2
            p2 = tmp

        return

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-9.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head: ListNode | null, n: number): ListNode {
        const dummy = new ListNode(-1, head)
        let curr: ListNode | null = head

        for (let i = 0; i < n; i++) {
            curr = curr.next
        }

        let prev = dummy
        while (curr) {
            curr = curr.next
            prev = prev.next
        }

        prev.next = prev.next.next

        return dummy.next
    }
}
